hack/generator/pkg/functions/original_gvk_function.go

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package functions
+
+import (
+	"github.com/dave/dst"
+
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astbuilder"
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
+)
+
+// OriginalGVKFunction implements a function to return the GVK for the type on which it is called
+// We put these on our resource types, giving us a way to obtain the right type of instance when the reconciler is
+// working with ARM. The code differs slightly depending on whether we're injecting into an API or storage variant.
+//
+// func (resource *SomeResource) OriginalGVK() scheme.GroupVersionKind {
+//     return scheme.GroupVersionKind{
+//         Group: GroupVersion.Group,
+//         Version: resource.Spec.OriginalVersion,
+//         Kind: "SomeResource",
+//     }
+// }
+//
+type OriginalGVKFunction struct {
+	idFactory                  astmodel.IdentifierFactory
+	hasOriginalVersionFunction bool
+	hasOriginalVersionProperty bool
+}
+
+// Ensure OriginalGVKFunction properly implements Function
+var _ astmodel.Function = &OriginalGVKFunction{}
+
+type OriginalVersionKind string
+
+const (
+	ReadProperty = OriginalVersionKind("property")
```

Comment:
I hate that go doesn't have first class enum support, and the solution of consts _kinda_ works but if the consts don't have the enum name in them it's sometimes hard to pick out which values belong to which type.

I know in some places we do something like `OriginalVersionKindReadProperty` (which is sorta how you'd do it in C, I think?), but we don't have that consistently and it is awkwardly long.

Does this bug other folks too? What do people prefer?

Reply:
I used very simple names here to optimize the API when consuming; adding a unique suffix would make that clunky. 

Having first class enum support would be preferable - but I guess Go is designed to push people towards many simpler packages.